### ISSUE TITLE
Emacs mode should use package.el

### ISSUE THREAD

**Issue by spwhitton** [agda/agda#2032]
Hello,

It would be nice if the Emacs mode was packaged as a package.el package.

Here is a sample `agda2-mode-pkg.el`:

```
(define-package "agda2-mode" "2.5.1"
  "interactive development for Agda, a dependently typed functional programming language")
```

Thanks.

**Comment by Blaisorblade** [comment]
Hi. First, thanks for the suggestion! Could you elaborate a bit?
1. Is that code sufficient? Where would it go (OK, probably https://github.com/agda/agda/tree/stable-2.5/src/data/emacs-mode), and how would one take advantage of it? Doesn't one need to also publish it through MELPA/Marmalade/...?
2. Unlike for other modes, the version of agda-mode must match exactly the version of Agda, which often must match closely the version of your Agda sources. How would this work with `package.el`? Also, what if I upgrade all my packages? Do I also have to upgrade Agda and have my Agda sources break? To fix this, some Emacs packages also ship associated programs (say, `ensime` and `intero`), but that also has issues.
3. Was `agda-mode setup` inconvenient to use? Maybe the issue can be addressed some other way?

**Comment by spwhitton** [comment]
Dear Paulo,

On Sun, Jun 12, 2016 at 05:42:30AM -0700, Paolo G. Giarrusso wrote:

> 1. Is that code sufficient? Where would it go (OK, probably
>    https://github.com /agda/agda/tree/stable-2.5/src/data/emacs-mode),
>    and how would one take advantage of it? Doesn't one need to also
>    publish it through MELPA/ Marmalade/...?

Yes: see [1].  A tarball containing all the .el files in that directory
constitutes a package.  You could choose to publish that package on
MELPA if you wanted to, but the use case I had in mind does not depend
on that publication -- see below.

> 1. Unlike for other modes, the version of agda-mode must match
>    exactly the version of Agda, which often must match closely the
>    version of your Agda sources. How would this work with package.el?
>    Also, what if I upgrade all my packages? Do I also have to upgrade
>    Agda and have my Agda sources break?  To fix this, some Emacs
>    packages also ship associated programs (say, ensime and intero), but
>    that also has issues.

As far as I know the Emacs packaging system has no mechanism to keep the
agda version in sync with the agda-mode version.  However, distribution
packages of agda-mode do: they can declare a tightly-versioned
dependency.

> 1. Was agda-mode setup inconvenient to use? Maybe the issue can be
>    addressed some other way?

My main motivation for writing the `agda2-mode-pkg.el` file I quoted
above was to standardise Debian's packaging of the Emacs mode.
Previously, the package just installed the .el files to a directory and
added some ad-hoc startup code to the `site-start` directory so that
agda-mode would be available for users to load.  However, with the
advent of package.el packages we can now install Emacs addons as
package.el packages, and make these available to a user's Emacs instance
in a unified way.

Since the user's Emacs is aware of agda-mode as a package rather than
just as a bunch of lisp spread out between a few files, some
functionality becomes available: for example, other package.el packages
can depend on (a new enough version of) agda-mode.  But I admit that the
main advantage is for distribution packagers rather than users.

Thanks.

[1] https://www.gnu.org/software/emacs/manual/html_node/elisp/Multi_002dfile-Packages.html

**Comment by Blaisorblade** [comment]
Ah, now I get your point, though not yet in full detail (I should study your link more, no time right now).

Would this hurt people using `agda-mode setup`, or alter that setup in any way?
Would one still need to bytecompile code by hand? Avoiding that would be nice.
From my POV, I guess one would need to ensure that setup instructions are up-to-date after the change.

I'll still need to defer to people more experienced on agda-mode — @nad is the last who did substantial work there?
But I do imagine that having a PR for this might help getting it merged.

**Comment by spwhitton** [comment]
On Mon, Jun 13, 2016 at 03:16:37AM -0700, Paolo G. Giarrusso wrote:

> Would this hurt people using agda-mode setup, or alter that setup in any way?
> Would one still need to bytecompile code by hand? Avoiding that would be nice.
> From my POV, I guess one would need to ensure that setup instructions are
> up-to-date after the change.

Ah, that's another advantage of packages :) I think that if you install
one into your home directory, Emacs automatically bytecompiles it.

> I'll still need to defer to people more experienced on agda-mode —
> @nad is the last who did substantial work there?  But I do imagine
> that having a PR for this might help getting it merged.

Done.

**Comment by spwhitton** [comment]
And no, I'm pretty sure it wouldn't prevent people installing with the
old method.

**Comment by Blaisorblade** [comment]
Given #2040 was merged, I'm happy to close this as fixed! (But let me know if I missed sth.)

**Comment by asr** [comment]
@Blaisorblade please add the 2.5.2 milestone.

**Comment by spwhitton** [comment]
This enables making an ELPA package out of your source code, so from
Debian's point of view that's all that's wanted.  You might consider
publishing the package on MELPA or something; that might be a reason to
keep the issue open.

**Comment by Blaisorblade** [comment]
@spwhitton Thanks for mentioning that, but without a solution to the issues I described, I'd rather avoid publishing the package.

**Comment by fredefox** [comment]
I have a problem that may (or may not) be related to this issue.

    § which agda-mode
    > ~/.local/bin/agda-mode
    § agda-mode --version
    > This program, which is part of Agda version 2.6.0, can be run
    ...

Yet when I try to load an Agda-file in Emacs I get the error:

    agda2-restart: The Agda mode's version (2.5.1) does not match that of agda (unknown).

The thing to note here is that the version of `agda-mode` doesn't match the one in my path &mdash; and of course that it can't determine the version of `agda` &mdash; which I really don't understand.

If I do `M-x agda2-set-program-version` and hit enter (not specifying a value when prompted) I can load the file without any problems. It's just really a bother to have to execute this command every time I start Emacs. I suspect something in Emacs has ended up in an inconsistent state, but I really can't track it down.

Any help would be greatly appreciated.

**Comment by fredefox** [comment]
I should maybe add that I don't know where this 2.5.1-version is coming from. I can't find it anywhere in my system xD

**Comment by andreasabel** [comment]
What is the output of `agda-mode --locate`?

**Comment by fredefox** [comment]
§ agda-mode locate  
    > /home/fredefox/workspace/agda/.stack-work/install/x86_64-linux-nopie/lts-7.15/8.0.1/share/x86_64-linux-ghc-8.0.1/Agda-2.6.0/emacs-mode/agda2.el

**Comment by fredefox** [comment]
I see a potential issue. As you may have gathered I've build agda from source using `stack`.

When you do `stack build` it puts the binaries in a location like the above, when you then do `stack install` it copies the binaries to `~/.local/bin` so perhaps the way agda detects where it lives interferes with this pattern. I dunno. Just a thought...

**Comment by fredefox** [comment]
So... I am tbh a total emacs-noob, but re-reading the comments above I decided to try this: comment out everything agda-related (`.spacemacs` -- spacemacs have a notion of configuration-layers. I don't fully understand how it works) and then in `init.el` I have (and had before-hand) the result of `agda-mode setup` i.e. the agda-mode locate thingy. It seems to be resolved now. Though I don't know if this is the ideal solution. At least I'm not getting this annoying error every time.

TL;DR; Was probably caused by `spacemacs`'s "configuration layers".

**Comment by Blaisorblade** [comment]
> So... I am tbh a total emacs-noob, but re-reading the comments above I decided to try this: comment out everything agda-related (.spacemacs -- spacemacs have a notion of configuration-layers. I don't fully understand how it works) and then in init.el I have (and had before-hand) the result of agda-mode setup i.e. the agda-mode locate thingy.

Any chance your .spacemacs config contained an out-of-date agda2.el (rather than a call to `agda-mode locate`?

Alternatively: Maybe unrelated (but surely related to this issue), `./src/data/emacs-mode/agda2-mode-pkg.el` still uses version 2.5.1 on both releases 2.5.2 and on master. That should affect packages (CC @spwhitton), and would affect any code relying on that file (possibly including spacemacs code).

**Comment by Blaisorblade** [comment]
I've manually fixed `agda2-mode-pkg.el` on `stable-2.5` and `master` (pushed directly). I hope that wasn't too eager.

I've added a PR for the release docs so this is avoided in the future, reopening till the PR is merged or some other fix is found.

**Comment by fredefox** [comment]
`.spacemacs` contained:

    dotspacemacs-configuration-layers '( ... agda ...)
    dotspacemacs-additional-packages '(agda2-mode)

I've removed this entirely and just have this in `.init.el`

    (load-file (let ((coding-system-for-read 'utf-8))
                    (shell-command-to-string "agda-mode locate")))

as per `agda-mode setup`. Like I said, not an emacs-expert, so I don't know if this is ideal. But it seems to work.

On the one hand I like letting Emacs deal with those dependencies, but on the other hand I rely on having a bleeding-edge version of Agda on my system. So that's somewhat of a conundrum.

**Comment by Blaisorblade** [comment]
@fredefox If it works I wouldn't worry too much — but I'm no Emacs/Spacemacs expert and you might want to ask the author of that layer if there's one.

Googling found two Spacemacs layers that seem to not have this problem—they seem to rely on `agda-mode` as usual:
https://github.com/syl20bnr/spacemacs/tree/master/layers/%2Blang/agda (official?)
as does this older (more minimal?) one: https://github.com/liamoc/spacemacs-agda (older?).
Status labels are according to https://twitter.com/acid2/status/605275126947553280.

**Comment by asr** [comment]
> As you may have gathered I've build agda from source using `stack`.

Did you install a previous version of Agda using Cabal?

**Comment by fredefox** [comment]
@Blaisorblade: Thanks for the input. You're right I should just be happy that it seems to be working.

@asr: No I intentionally stay far away from using cabal directly. It's caused me many a headache.
